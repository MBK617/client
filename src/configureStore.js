/**
 * @file Configure redux store
 */

import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import createReducer from './reducers';

/**
 * Configures and returns initial redux store
 *
 * @param {object} initialState The initial state of the application
 * 
 * @return {object} A redux store
 * 
 * @memberof module:Main
 *
 */
function configureStore(initialState = {}) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    createReducer(),
    initialState,
    applyMiddleware(sagaMiddleware)
  );

  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {};
  store.injectedSagas = {};

  return store;
}

export default configureStore;