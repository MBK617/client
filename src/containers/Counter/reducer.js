import { fromJS } from 'immutable';

const initialState = fromJS({
  value : 0,
  incrementingAsynchronously: false
});

export default function counter(state = initialState, action) {
  switch (action.type) {
    case 'INCREMENT':
      return state.set('value', state.get('value') + 1)
    case 'DECREMENT':
      return state.set('value', state.get('value') - 1)
    case 'INCREMENT_ASYNC':
      return state.set('incrementingAsynchronously', true)
    case 'INCREMENT_ASYNC_SUCCESS':
      return state.set('incrementingAsynchronously', false)
    default:
      return state
  }
}
