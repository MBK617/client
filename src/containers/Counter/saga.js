import { increment, incrementAsyncSuccess } from "./actions";
import { delay, put, takeLatest, all } from 'redux-saga/effects';

export function* incrementAsyncSaga() {
  yield delay(1000)
  yield put(increment())
  yield put(incrementAsyncSuccess())
}

export default function* rootSaga() {
  yield all([
    takeLatest('INCREMENT_ASYNC', incrementAsyncSaga),
  ]);
}