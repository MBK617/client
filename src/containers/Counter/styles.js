import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  button: {
    margin: 12,
  },
  counter: {
    fontSize: 80,
    textAlign: 'center'
  }
});

export default styles;