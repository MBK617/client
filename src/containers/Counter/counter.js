import React from 'react';
import { compose } from 'redux';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';

import { 
  Button as ReactNativeButton, 
  View, 
  Text 
} from 'react-native';

import injectSaga from 'utils/redux/injectSaga';
import injectReducer from 'utils/redux/injectReducer';

import styles from './styles';
import reducer from './reducer';
import saga from './saga';
import { 
  increment, 
  decrement, 
  incrementAsync 
} from './actions';


const selectCounter = createSelector(
  state => state.counter,
  substate => substate.toJS()
);

const Button = (props) => (
  <View style={styles.button}>
    <ReactNativeButton {...props} />
  </View>
);

const Counter = () => {
  const dispatch = useDispatch();
  const { value, incrementingAsynchronously: isLoading } = useSelector(selectCounter);

  return (
    <View>
      <Text style={styles.counter}>
        {value}
      </Text>
      <Button 
        onPress={()=>dispatch(incrementAsync())}
        title={`Increment after 1 second ${isLoading ? '...' : ''}`}
        disabled={isLoading}
      />
      <Button 
        onPress={()=>dispatch(increment())}
        title='Increment'
      />
      <Button 
        onPress={()=>dispatch(decrement())}
        title='Decrement'
      />
    </View>
  );
}

const withReducer = injectReducer({ key: 'counter', reducer });
const withSaga = injectSaga({ key: 'counter', saga });

export default compose(withReducer, withSaga)(Counter);
