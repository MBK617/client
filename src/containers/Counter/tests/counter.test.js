import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render, cleanup, fireEvent } from '@testing-library/react-native';
import { fromJS } from 'immutable';
import createSagaMiddleware from 'redux-saga';


import Counter from '../counter';


const initialState = {
  counter: fromJS({
    value : 0,
    incrementingAsynchronously: false
  })
};



describe('<Counter />', () => {
  const sagaMiddleware = createSagaMiddleware();
  let store, rendered;

  const mockStore = (initialState) => ({ 
    ...configureStore([sagaMiddleware])(initialState), 
    injectedReducers: {}, 
    injectedSagas: {}, 
    runSaga: sagaMiddleware.run 
  });

  beforeEach( () => {
    store = mockStore(initialState);
    rendered = render(
      <Provider store={store}><Counter /></Provider>
    );
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display current count', () => {
    const textComponent = rendered.getByText('0');
    expect(textComponent.props.children).toEqual(0);
  });
  
  it('should dispatch increment action', () => {
    const buttonComponent = rendered.getByText('Increment');
    fireEvent(buttonComponent, 'press');
    const actions = store.getActions();
    expect(actions.length).toBe(1);
    expect(actions[0].type).toEqual('INCREMENT');
  });

  it('should dispatch decrement action', () => {
    const buttonComponent = rendered.getByText('Decrement');
    fireEvent(buttonComponent, 'press');
    const actions = store.getActions();
    expect(actions.length).toBe(1);
    expect(actions[0].type).toEqual('DECREMENT');
  });

  it('should dispatch increment async action', () => {
    const buttonComponent = rendered.getByText('Increment after 1 second ');
    fireEvent(buttonComponent, 'press');
    const actions = store.getActions();
    expect(actions.length).toBe(1);
    expect(actions[0].type).toEqual('INCREMENT_ASYNC');
  });

});