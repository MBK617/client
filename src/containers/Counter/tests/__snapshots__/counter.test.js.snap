// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> should render correctly 1`] = `
<View>
  <Text
    style={
      Object {
        "fontSize": 80,
        "textAlign": "center",
      }
    }
  >
    0
  </Text>
  <View
    style={
      Object {
        "margin": 12,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          disabled={false}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Increment after 1 second 
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "margin": 12,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Increment
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "margin": 12,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Decrement
        </Text>
      </View>
    </View>
  </View>
</View>
`;
